coderd/workspacebuilds.go

```diff
@@ -327,6 +328,16 @@ func (api *API) postWorkspaceBuilds(rw http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	previousWorkspaceBuild, err := api.Database.GetLatestWorkspaceBuildByWorkspaceID(ctx, workspace.ID)
```

Comment:
This is racy since we're not in a transaction. There could be an intermediate build between this one and the actual build (unlikely).

This and `builder.Build` could be moved closer to each other and wrapped inside a tx to eliminate the edge case, though (the same is done in `createWorkspace` so should be fine). Or since the builder already fetches the previous build, perhaps the deduction of "template changed" can be moved in there and exposed by some means.

(Not looking to complicate this so feel free to ignore, but yet another alternative is to move the notification logic to wsbuilder, although that may need to rely on initiator vs the signal of using this endpoint. The initiator could be you, or it could be an admin, etc.)

Reply:
I think the likelihood of a race is very minimal, but even if it should happen the notifier would handle de-duplication anyways.

I also think moving the notification logic into `wsbuilder` is at the wrong abstraction level.